crates/sui-indexer/src/restorer/formal_snapshot.rs

```diff
@@ -0,0 +1,231 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::collections::BTreeMap;
+use std::fs;
+use std::num::NonZeroUsize;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+use futures::future::{AbortHandle, AbortRegistration, Abortable};
+use futures::{StreamExt, TryStreamExt};
+use indicatif::{MultiProgress, ProgressBar, ProgressStyle};
+use object_store::path::Path;
+use tokio::sync::Mutex;
+use tracing::info;
+
+use sui_config::object_storage_config::{ObjectStoreConfig, ObjectStoreType};
+use sui_core::authority::authority_store_tables::LiveObject;
+use sui_snapshot::reader::{download_bytes, LiveObjectIter, StateSnapshotReaderV1};
+use sui_snapshot::FileMetadata;
+use sui_storage::object_store::ObjectStoreGetExt;
+use sui_types::accumulator::Accumulator;
+
+use crate::config::RestoreConfig;
+use crate::errors::IndexerError;
+use crate::handlers::TransactionObjectChangesToCommit;
+use crate::restorer::archives::read_next_checkpoint_after_epoch;
+use crate::store::{indexer_store::IndexerStore, PgIndexerStore};
+use crate::types::{IndexedDeletedObject, IndexedObject};
+
+pub type DigestByBucketAndPartition = BTreeMap<u32, BTreeMap<u32, [u8; 32]>>;
+pub type SnapshotChecksums = (DigestByBucketAndPartition, Accumulator);
+pub type Sha3DigestType = Arc<Mutex<BTreeMap<u32, BTreeMap<u32, [u8; 32]>>>>;
+
+pub struct IndexerFormalSnapshotRestorer {
+    store: PgIndexerStore,
+    reader: StateSnapshotReaderV1,
+    restore_config: RestoreConfig,
+    next_checkpoint_after_epoch: u64,
+}
+
+impl IndexerFormalSnapshotRestorer {
+    pub async fn new(
+        store: PgIndexerStore,
+        restore_config: RestoreConfig,
+    ) -> Result<Self, IndexerError> {
+        let remote_store_config = ObjectStoreConfig {
+            object_store: Some(ObjectStoreType::S3),
+            aws_endpoint: Some(restore_config.s3_endpoint.clone()),
+            aws_virtual_hosted_style_request: true,
+            object_store_connection_limit: restore_config.object_store_concurrent_limit,
+            no_sign_request: true,
+            ..Default::default()
+        };
+
+        let base_path = PathBuf::from(restore_config.gcs_snapshot_dir.clone());
+        let snapshot_dir = base_path.join("snapshot");
+        if snapshot_dir.exists() {
+            fs::remove_dir_all(snapshot_dir.clone()).unwrap();
+            info!(
+                "Deleted all files from snapshot directory: {:?}",
+                snapshot_dir
+            );
+        } else {
+            fs::create_dir(snapshot_dir.clone()).unwrap();
+            info!("Created snapshot directory: {:?}", snapshot_dir);
+        }
```

Comment:
This is the directory where the formal snapshot is downloaded to? If so we shouldn't nuke it if it exists. We want this process to be restartable in the event that the snapshot download or restore to PG was interrupted.

Reply:
yes but to support resuming from partial download, `StateSnapshotReaderV1` needs to support partial download, which is not there yet. Also today the downloading part only takes less than 10 min, so maybe a nice-to-have later on but not urgent today if that makes sense.